bufferserver/src/main/java/com/datatorrent/bufferserver/internal/DataList.java

```diff
@@ -862,10 +902,23 @@ private boolean canEvict(final int refCount, boolean writer)
         } else {
           return true;
         }
-      } else {
-        logger.debug("Holding {} in memory due to {} references.", this, refCount);
-        return false;
       }
+      logger.debug("Not evicting {} due to {} references.", this, refCount);
+      return false;
+    }
+
+    private boolean isPublisherAheadByMax()
+    {
+      boolean ahead = false;
```

Comment:
As release may be called when Subscriber disconnects and it's iterator is closed, it is possible that it was the last Subscriber that holds the Publisher. It is necessary to do 2 stage processing similar to what is done in areSubscribersBehindByMax: find the first block with positive reference count and see if Publisher is max block ahead of it.